README.rst
service-checker documentation
=============================

A generic swagger-based webservice checker.

It can perform http requests based on the ``x-amples`` sections of the
``paths`` part of the spec, and match responses to what the expected
response is. It is able to check both headers and the body of the response.

Installation
------------

This software is known to work correctly with python 2.7 and 3.5

From Source
~~~~~~~~~~~

.. code:: bash

    $ python setup.py install

Usage
-----

Once installed, you will have the ``service-checker-swagger`` binary
in your path.

Suppose you have a webservice running on localhost port 8080, which
reponds for HTTP host ``awesomeservice.local`` and it
exposes its swagger spec on the /swagger url. To check it is working
as designed according to its spec you can just do

.. code:: bash

    $ service-checker-swagger 127.0.0.1 awesomeservice.local:8080 -s /swagger
    All endpoints are healthy

service-checker-swagger supports HTTPS urls just fine, but in some cases, you
might want to avoid certificate verification. While this is something you
should not do unless you understand very very well the reasons for it, it is
supported via the --insecure (-k shorthand)

.. code:: bash

    $ service-checker-swagger -k 127.0.0.1 https://awesomeservice.local:8080 -s /swagger
    All endpoints are healthy


Spec format support
-------------------

``service-checker-swagger`` checks each of the paths in your swagger/OpenAPI
specification for an ``x-amples`` section and uses it to do live requests
to the API and checks that the response corresponds to the base. The
``x-amples`` section is an extension to the swagger spec that has been
introduced by `swagger-test <https://github.com/earldouglas/swagger-test>`_
and that consists of items with ``request`` and ``response`` sections.

Url templating according to RFC 6570 is partially supported via the
``params`` section of the ``x-amples`` section.

There could be some specific examples you might want to run unit tests
on but you don't want to monitor on a live service (typically, any
non-idempotent request is a good candidate for this). In that case,
just add ``x-monitor: false`` at the root of your example.

Basic example
~~~~~~~~~~~~~
.. code:: javascript

   "/pets/{id}": {
       "get": {
          "x-monitor": true,
          "x-amples": [ {
            "request": {
              "params": {"id": 10},
              "headers": { "Accept": "application/json", },
              "query": {"refresh": "y"},
            },
            "response": {
              "status": 200,
              "headers": {"X-Pet-Iscute": "yes"},
              "body": { "species": "/canis .*/"}
            },
            ...

Url template interpolation
~~~~~~~~~~~~~~~~~~~~~~~~~~

We support basic Url template interpolation, a subset of the
specification in RFC 6570 is supported at the moment. We support
simple, optional and multiple parameter substitutions.



Body data check
~~~~~~~~~~~~~~~

Body data is assumed to be either json or plain text; actual content
(of either fields in the json structure or the text) can be either
matched exactly or with a regexp. So for example:

.. code:: javascript

    "body": "abcd"

will verify that "abcd" is the exact response body, while

.. code:: javascript

    "body": "/^abcd/"

will just check that the body begins with "abcd".

Limitations
-----------

- Only supports GET and POST at the moment
- Only plain-text and json responses are supported
- Url templating support is pretty limited at the moment
- All endpoints are checked sequentially, which could easily lead to
  timeouts in nagios-like systems
- No logging


debian/changelog
service-checker (0.2.1-1) stretch-wikimedia; urgency=medium

  * Allow skipping cert verification
  * Drop trusty support
  * Drop python2, python 3.4 support
  * Add python3.7 (which is in Debian Buster) support
  * Fix various debianization issues

 -- Alexandros Kosiaris <akosiaris@wikimedia.org>  Tue, 12 Jan 2021 15:23:38 +0200

service-checker (0.2.0-1) stretch-wikimedia; urgency=medium

  * Add logging
  * Add generic interface to metrics gathering
  * Extend dpkg-source diff ignore
  * Copy fixtures while building the package
  * Only build for python3

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Wed, 18 Sep 2019 12:17:45 +0200

service-checker (0.1.5-1) jessie-wikimedia; urgency=medium

  * Support OpenAPI v3 specifications
  * Make raw octet-stream requests properly

 -- Marko Obrovac <mobrovac@wikimedia.org>  Wed, 1 May 2019 02:30:00 +0000

service-checker (0.1.4-2) unstable; urgency=medium

  * Adding runtime dependency to pkg_resources which is used by the cli script

 -- Alexandros Kosiaris <alex@wikimedia.org>  Fri, 09 Feb 2018 18:46:15 +0200

service-checker (0.1.4-1) unstable; urgency=medium

  * Conversion to a non-native Debian package

 -- Alexandros Kosiaris <alex@wikimedia.org>  Tue, 06 Feb 2018 15:07:36 +0200

service-checker (0.1.4) jessie-wikimedia; urgency=medium

  * Improve the checking procedure and emit better messages

 -- Marko Obrovac <mobrovac@wikimedia.org>  Thu, 26 Oct 2017 13:30:00 +0000

service-checker (0.1.3) jessie-wikimedia; urgency=medium

  * Fixup for non-https urls

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Fri, 09 Jun 2017 17:34:43 +0200

service-checker (0.1.2) jessie-wikimedia; urgency=low

  [ Marko Obrovac ]
  * Include the User-Agent header field when making requests

  [ Giuseppe Lavagetto ]
  * Add 'assert_username' to PoolManager to better handle TLS.

 -- Marko Obrovac <mobrovac@wikimedia.org>  Tue, 06 Jun 2017 14:02:10 +0000

service-checker (0.1.1) UNRELEASED; urgency=medium

  * Fix double-inclusion of prefix

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Mon, 08 May 2017 17:58:10 +0200

service-checker (0.1.0) jessie-wikimedia; urgency=medium

  * Asyncronous url fetching using gevent
  * Added ability to report timing data to statsd

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Fri, 05 May 2017 13:26:53 +0200

service-checker (0.0.3) jessie-wikimedia; urgency=medium

  * Improved message on malformed responses

 -- Riccardo Coccioli <rcoccioli@wikimedia.org>  Tue, 29 Nov 2016 12:11:52 +0100

service-checker (0.0.2) jessie-wikimedia; urgency=medium

  * Add support for x-default-query
  * Can now be used for checking MediaWiki
  * Supports yaml specs

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Fri, 28 Oct 2016 12:25:24 +0200

service-checker (0.0.1) unstable; urgency=medium

  * Initial release

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Tue, 05 Jul 2016 17:51:24 +0200


checker/tests/unit/test_checker.py
import checker
from checker import service
import unittest
import mock
import json
import os
import urllib3
import copy


class TestTemplateUrl(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.t = service.TemplateUrl(
            'https://example.org/test/{where}{/what}{/+why}')

    def test_init(self):
        """
        Test initialization of the template url
        """
        origs = [el.original for el in self.t.tokens]
        self.assertEquals(origs, ['{where}', '{/what}', '{/+why}'])
        self.assertEquals(self.t.tokens[0].types, ['simple'])
        self.assertEquals(self.t.tokens[1].types, ['simple', 'optional'])

    def test_realize(self):
        """
        Test Realization
        """
        params = {'where': 'Rome', 'what': 'Eat', 'why': ['I', 'am', 'hungry']}
        self.assertEquals(self.t.realize(params),
                          'https://example.org/test/Rome/Eat/I/am/hungry')

        params['what'] = 'Eat:pasta'
        self.assertEquals(
            self.t.realize(params),
            'https://example.org/test/Rome/Eat%3Apasta/I/am/hungry')
        params1 = {'where': 'Rome'}
        self.assertEquals(self.t.realize(params1),
                          'https://example.org/test/Rome')

        del params['why']
        self.assertEquals(self.t.realize(params),
                          'https://example.org/test/Rome/Eat%3Apasta')


class TestEndpointRequest(unittest.TestCase):

    def mock_response(self, d):
        r = mock.create_autospec(urllib3.response.HTTPResponse)
        r.status = d['status']
        r.data = json.dumps(d['body']).encode('utf-8')

        def getheader(key):
            return d['headers'].get(key, None)
        r.getheader = mock.MagicMock(side_effect=getheader)
        service.fetch_url = mock.MagicMock(return_value=r)

    def setUp(self):
        self.resp = {
            'body': {
                'has_items': True,
                'items': [
                    {
                        'comment': '/.*/',
                        'rev': '/\\d+/',
                        'tid': '/^[0-9a-fA-F]{4}-[0-9a-fA-F]{4}$/',
                        'title': 'Foobar'
                    }
                ]
            },
            'headers': {
                'content-type': 'application/json',
                'etag': '/.+/'
            },
            'status': 200
        }

        self.ep = service.EndpointRequest(
            "a Test endpoint",
            "http://127.0.0.1/baseurl",
            "get",
            "/an/endpoint/{revision}",
            {"http_host": 'example.org', 'params': {'revision': 1}},
            copy.deepcopy(self.resp)
        )
        self.resp["body"]["items"] = [
            {"title": "Foobar",
             "comment": "blabla",
             "rev": 1,
             "tid": "00AA-abcd"}
        ]
        self.resp["body"]["has_items"] = True
        self.resp["headers"]["etag"] = "Imtrackingyou"

    def test_init(self):
        """
        Test initialization
        """
        self.assertEquals(self.ep.title, "a Test endpoint")
        self.assertEquals(self.ep.tpl_url._url_string,
                          "http://127.0.0.1/baseurl/an/endpoint/{revision}")
        self.assertEquals(self.ep.request_headers, {'Host': 'example.org'})
        self.assertEquals(self.ep.url_parameters, {'revision': 1})
        self.assertEquals(self.ep.resp_status, 200)
        self.assertTrue(self.ep.headers["content-type"]('application/json'))

    def test_run_ok(self):
        """
        Test a successful run
        """
        self.mock_response(self.resp)
        self.ep.run(urllib3.PoolManager())
        self.assertEquals(self.ep.status, 'OK')

    def test_run_bad_status(self):
        """
        Test an unexpected HTTP status
        """
        self.resp['status'] = 301
        self.mock_response(self.resp)
        self.ep.run(urllib3.PoolManager())
        self.assertEquals(self.ep.status, 'CRITICAL')
        self.assertEquals("Test a Test endpoint returned "
                          "the unexpected status 301 (expecting: 200)",
                          self.ep.msg)

    def test_run_bad_header(self):
        """
        Test an unexpected HTTP Header
        """
        self.resp['headers']['etag'] = ""
        self.mock_response(self.resp)
        self.ep.run(urllib3.PoolManager())
        self.assertEquals(self.ep.status, 'CRITICAL')
        self.assertEquals("Test a Test endpoint had an unexpected value "
                          "for header etag: ", self.ep.msg)

    def test_run_missing_header(self):
        """
        Test a missing HTTP header
        """
        del self.resp['headers']['etag']
        self.mock_response(self.resp)
        self.ep.run(urllib3.PoolManager())
        self.assertEquals(self.ep.status, 'CRITICAL')
        self.assertEquals("Test a Test endpoint had an unexpected value "
                          "for header etag: None", self.ep.msg)

    def test_run_bad_body(self):
        """
        Test unexpected value in body
        """
        self.resp['body']['items'][0]['tid'] = 12
        self.mock_response(self.resp)
        self.ep.run(urllib3.PoolManager())
        self.assertEquals(self.ep.status, 'WARNING')
        self.assertEquals("Test a Test endpoint responds with unexpected "
                          "body: /items[0]/tid => 12", self.ep.msg)

    def test_run_missing_body(self):
        """
        Test missing value in body
        """
        del self.resp['body']['items'][0]['tid']
        self.mock_response(self.resp)
        self.ep.run(urllib3.PoolManager())
        self.assertEquals(self.ep.status, 'WARNING')
        self.assertEquals("Test a Test endpoint responds with unexpected "
                          "body: /items[0]/tid => None", self.ep.msg)

    def test_default_response(self):
        """
        Test a simple endpoint
        """
        ep = service.EndpointRequest(
            "simple test",
            "http://127.0.0.1:7321",
            "get",
            "/test",
            {'http_host': "example.org"},
            {'status': 200},
        )
        self.mock_response(
            {"status": 200,
             "body": "Hello, World!",
             "headers": {"content-length": 3240}})
        ep.run(urllib3.PoolManager())
        self.assertEquals(self.ep.status, 'OK')


class TestCheckService(unittest.TestCase):
    routes = {}

    def add_mock_response(self, route, d):
        r = mock.create_autospec(urllib3.response.HTTPResponse)
        r.status = d['status']
        r.data = json.dumps(d['body']).encode('utf-8')

        def getheader(key):
            return d['headers'].get(key, None)
        r.getheader = mock.MagicMock(side_effect=getheader)
        self.routes[route] = r

    def router(self, client, route, **kw):
        r = route.replace(self.cs._url, '')
        return self.routes.get(r, None)

    def mock_routes(self):
        service.fetch_url = mock.MagicMock(side_effect=self.router)

    def setUp(self):
        self.cs = service.CheckService('127.0.0.1', 'http://example.org/api')
        fn = os.path.join(os.path.dirname(__file__), '../fixtures/test.json')
        with open(fn, 'rb') as f:
            data = f.read().decode('utf-8')
        self.add_mock_response(
            '/?spec', {'status': 200, 'body': json.loads(data)})

    def test_initialize(self):
        """
        Test initialization
        """
        self.assertEquals(self.cs.host_ip, '127.0.0.1')
        self.assertEquals(self.cs._timeout, 5)
        self.assertEquals(self.cs.port, '80')
        self.assertEquals(self.cs.http_host, 'example.org')
        self.assertEquals(self.cs._url, 'http://127.0.0.1:80/api')

    def test_get_endpoints(self):
        """
        Test list of endpoints is returned
        """
        self.mock_routes()
        l = [el for el, data in self.cs.get_endpoints()]
        l.sort()
        self.assertListEqual(l, [u'/simple', u'/{who}/{verb}'])

    def test_get_ep_invalid_spec(self):
        """
        Test what endpoints are returned with incorrect specs
        """
        fn = os.path.join(os.path.dirname(__file__), '../fixtures/test_error_spec.json')
        with open(fn, 'rb') as f:
            data = f.read().decode('utf-8')
        self.add_mock_response(
            '/?spec', {'status': 200, 'body': json.loads(data)})
        self.mock_routes()
        l = [el for el, _ in self.cs.get_endpoints()]
        self.assertEquals(l, [u'/{who}/{verb}'])

    def test_run(self):
        """
        Test a successful run
        """
        self.add_mock_response('/simple', {'status': 200, 'body': 'hi'})
        self.add_mock_response(
            '/joe/rulez', {'status': 200, 'body': 'For sure!'})
        self.mock_routes()
        with self.assertRaises(SystemExit) as e:
            self.cs.run()
        self.assertEquals(e.exception.code, 0)

    def test_endpoint_critical(self):
        """
        Test a critical exit
        """
        self.add_mock_response('/simple', {'status': 200, 'body': 'hi'})
        self.add_mock_response('/joe/rulez', {'status': 301, 'body': ''})
        self.mock_routes()
        with self.assertRaises(SystemExit) as e:
            self.cs.run()
        self.assertEquals(e.exception.code, 2)

    def test_endpoint_warning(self):
        """
        Test a warning exit
        """
        self.add_mock_response('/simple', {'status': 200, 'body': 'hi'})
        self.add_mock_response(
            '/joe/rulez', {'status': 200, 'body': 'For sure?'})
        self.mock_routes()
        with self.assertRaises(SystemExit) as e:
            self.cs.run()
        self.assertEquals(e.exception.code, 1)


servicechecker/swagger.py
#!/usr/bin/env python
# Import gevent.monkey and do the monkey patching
from gevent import monkey; monkey.patch_all()  # noqa

import argparse
from collections import namedtuple
import json
import yaml
import re
import sys
import os

import gevent

# python 2 vs python 3 imports
try:
    import urlparse
    from urllib import quote_plus
except ImportError:
    import urllib.parse as urlparse
    from urllib.parse import quote_plus

try:
    reload(sys)  # noqa
    sys.setdefaultencoding('utf-8')
except:
    pass

from servicechecker import CheckerBase, fetch_url, CheckError, logging
from servicechecker.metrics import Null as NullMetrics
from servicechecker.metrics import StatsD as StatsDMetrics


class CheckService(CheckerBase):
    """
    Shell class for checking services
    """
    default_request = {
        'request': {},
        'response': {'status': 200},
        'title': 'Untitled test'
    }
    _supported_methods = ['get', 'post']

    def __init__(self, host_ip, base_url, timeout=5, spec_url='/?spec',
                 metrics_manager=None, insecure=False):
        """
        Initialize the checker

        Args:
            host_ip (str): The host ipv4 address (also works with a hostname)

            base_url (str): The base url the service expects to respond from

            timeout (int): Number of seconds to wait for each request

            spec_url (str): the string to append to the base url, defaults to
                           /?spec
           metrics_manager (Metrics): Instance of Metrics implementation or None. default: None
        """
        self.host_ip = host_ip
        self.base_url = urlparse.urlsplit(base_url)
        http_host_port = self.base_url.netloc.split(':')
        if len(http_host_port) < 2:
            if self.base_url.scheme == 'https':
                http_host_port.append('443')
            else:
                http_host_port.append('80')
        self.http_host, self.port = http_host_port
        self._url_prefix = self.base_url.path
        self.endpoints = {}
        self._timeout = timeout
        self.spec_url = spec_url
        self.is_https = (self.base_url.scheme == 'https')
        self.insecure = insecure
        if metrics_manager is None:
            self.metrics_manager = NullMetrics()
        else:
            self.metrics_manager = metrics_manager

    @property
    def _url(self):
        """
        Returns an url pointing to the IP of the host to check.
        """
        return "{}://{}:{}".format(self.base_url.scheme,
                                   self.host_ip,
                                   self.port)

    def get_endpoints(self):
        """
        Gets the full spec from base_url + '/?spec' and parses it.
        Returns a generator iterating over the available endpoints
        """
        http = self._spawn_downloader(self.is_https, self.insecure)
        # TODO: cache all this.
        response = fetch_url(
            http,
            self._url + self._url_prefix + self.spec_url,
            timeout=self._timeout,
            headers={'Host': self.http_host}
        )

        resp = response.data.decode('utf-8')

        try:
            r = json.loads(resp)
        except ValueError:
            # try to load as YAML
            try:
                r = yaml.safe_load(resp)
            except yaml.YAMLError:
                raise ValueError("No valid spec found")

        # default params for URL interpolation
        TemplateUrl.default = r.get('x-default-params', {})
        # default query parameters for requests
        default_query = r.get('x-default-query', {})
        servers = r.get('servers', [])
        base_path = servers[0].get('url', '') if servers and servers[0] else r.get('basePath', '')
        base_path = base_path.rstrip('/')
        for endpoint, data in r['paths'].items():
            if not endpoint:
                continue
            endpoint = base_path + endpoint
            for key in self._supported_methods:
                try:
                    d = data[key]
                    # If x-monitor is False, skip this
                    if not d.get('x-monitor', True):
                        logging.debug('x-monitor is False, skipping %s %s' % (key, endpoint))
                        continue
                    if key == 'get':
                        default_example = [self.default_request.copy()]
                    else:
                        # Only GETs have default examples
                        default_example = []
                    examples = d.get('x-amples', default_example)
                    for x in examples:
                        req = {'http_method': key}
                        req.update(self.default_request)
                        req.update(x)
                        # Merge query parameters with defaults
                        # In Py 3.5 we could do {**default_query, **query}
                        query = default_query.copy()
                        query.update(req['request'].get('query', {}))
                        req['request']['query'] = query
                        yield endpoint, req
                except KeyError:
                    # No data for this method
                    pass

    def run(self):
        """
        Runs the checks on all the endpoints we find
        """
        res = []
        status = 'OK'
        idx = self.nagios_codes.index(status)
        # Spawn the downloaders
        checks = [{'ep': ep, 'data': data, 'job': gevent.spawn(self._check_endpoint, ep, data)}
                  for ep, data in self.get_endpoints()]
        # The -2 is for terminating the connections before the nrpe timeout
        # kicks in
        gevent.joinall([v['job'] for v in checks], self.nrpe_timeout - 2)

        for v in checks:
            endpoint = v['ep']
            data = v['data']
            title = data.get('title', "test for {}".format(endpoint))
            job = v['job']
            # Endpoint fetching failed or timed out.
            if not job.successful():
                status = 'CRITICAL'
                idx = self.nagios_codes.index(status)
                if job.exception:
                    res.append("{ep} - generic error: {exc}".format(ep=endpoint, exc=job.exception))
                else:
                    res.append(
                        '{ep} ({title}) timed out before a response was received'.format(
                            ep=endpoint, title=title,
                        )
                    )
            else:

                ep_status, msg = job.value
                # WARNING or UNKNOWN
                if ep_status != 'OK':
                    res.append(
                        "{ep} ({title}) is {status}: {message}".format(
                            ep=endpoint, title=title, status=ep_status,
                            message=msg
                        )
                    )
                    ep_idx = self.nagios_codes.index(ep_status)
                    if ep_idx >= idx:
                        status = ep_status
                        idx = ep_idx

        if status == 'OK':
            message = "All endpoints are healthy"
        else:
            message = u"; ".join(res)
        print(message)
        sys.exit(idx)

    def _check_endpoint(self, endpoint, data):
        """
        Actually performs the checks on each single endpoint
        """
        req = data.get('request', {})
        req['http_host'] = self.http_host
        er = EndpointRequest(
            data.get('title',
                     "test for {}".format(endpoint)),
            self._url,
            data['http_method'],
            endpoint,
            req,
            data.get('response'),
            self.metrics_manager
        )
        er.run(self._spawn_downloader(self.is_https, self.insecure))
        return (er.status, er.msg)


class EndpointRequest(object):

    """
    Manages a request to a specific endpoint
    """

    def __init__(self, title, base_url, http_method,
                 endpoint,  request, response, metrics_manager=None):
        """
        Initialize the endpoint request

        Args:
            title (str): a descriptive name

            base_url (str): the base url

            http_method(str): the HTTP method

            endpoint (str): an url template for the endpoint, per RFC 6570

            request (dict): All data for building the request

            response (dict): What we should test in the response

            metrics_manager (Metrics): Instance of Metrics implementation or None. default: None
        """
        self.status = 'OK'
        self.msg = 'Test "{}" healthy'.format(title)
        self.title = title
        self.method = http_method
        self._request(request)
        self._response(response)
        self.tpl_url = TemplateUrl(base_url + endpoint)
        self.base_url = base_url
        if metrics_manager is None:
            self.metrics_manager = NullMetrics()
        else:
            self.metrics_manager = metrics_manager

    def run(self, client):
        """
        Perform the request, and test the result

        Args:
            client (urllib3.Poolmanager): the HTTP client we want to use
        """
        try:
            url = self.tpl_url.realize(self.url_parameters)
            with self.metrics_manager.record(url):
                r = fetch_url(
                    client,
                    url,
                    headers=self.request_headers,
                    fields=self.query_parameters,
                    redirect=False,
                    method=self.method
                )
        except CheckError as e:
            self.status = 'CRITICAL'
            self.msg = "Could not fetch url {}: {}".format(
                url, e)
            return

        # Response status
        if r.status != self.resp_status:
            self.status = "CRITICAL"
            self.msg = ("Test {} returned "
                        "the unexpected status {} (expecting: {})".format(
                            self.title, r.status, self.resp_status))
            return

        # Headers
        for k, v in self.headers.items():
            h = r.getheader(k)
            if h is None or not v(h):
                self.status = "CRITICAL"
                self.msg = ("Test {} had an unexpected value "
                            "for header {}: {}".format(self.title, k, h))
                return
        # Body
        if self.body is not None:
            body = r.data.decode('utf-8')
            if isinstance(self.body, dict) or isinstance(self.body, list):
                data = json.loads(body)
                try:
                    self._check_json_chunk(data, self.body)
                except CheckError:
                    return
                except Exception as e:
                    self.status = "CRITICAL"
                    self.msg = ("Test {} responds with malformed "
                                "body ({}: {}):\n{}".format(
                                    self.title,
                                    e.__class__.__name__,
                                    e.message,
                                    data))
            else:
                check = self._verify(self.body)
                if not check(body):
                    self.status = "WARNING"
                    self.msg = ("Test {} responds with unexpected "
                                "body: {} != {}".format(
                                    self.title,
                                    body,
                                    self.body))
                    return

    def _request(self, data):
        """
        Gather data from the request object
        """
        self.request_headers = {'Host': data['http_host']}
        if 'headers' in data:
            self.request_headers.update(data['headers'])
        self.url_parameters = data.get('params', {})
        qkey = 'query' if self.method == 'get' else 'body'
        self.query_parameters = data.get(qkey, {})

    def _response(self, data):
        """
        Organize the expected response data
        """
        self.resp_status = data['status']
        self.body = data.get('body', None)
        self.headers = {}
        try:
            for k, v in data['headers'].items():
                self.headers[k] = self._verify(v)
        except KeyError:
            pass

    def _verify(self, orig):
        """
        Return a lambda function to verify the response data

        Args:
            arg (str): The argument to check against. If enclosed
                       in slashes, it's assumed to be a regex
        """
        arg = str(orig)
        t = 'eq'
        if arg.startswith('/') and arg.endswith('/'):
            arg = arg.strip('/')
            t = 're'
        if t == 'eq':
            return lambda x: (x == arg) or x.startswith(arg)
        elif t == 're':
            return lambda x: re.search(arg, x)

    def _set_warning(self, prefix, data):
        """
        Sets self.status and self.msg to WARNING with
        an appropriate message, based on the args. If
        the status is set to something other than OK or
        WARNING, it does not update the state, so as
        not to potentially overwrite a CRITICAL message.

        Args:
            prefix (str): the body path being chcked

            data (mixed): the data to use in the message
        """
        if self.status not in ['OK', 'WARNING']:
            return False
        if self.status == 'OK':
            self.msg = ''
        self.status = "WARNING"
        self.msg += ("Test {} responds with unexpected "
                     "value at path {} => {}\n".format(self.title, prefix, data))
        return True

    def _check_json_chunk(self, data, model, prefix=''):
        """
        Recursively check a json chunk of the response.

        Args:
            data (mixed): the data to check

            model (mixed): the model to check the data against

            prefix (str): the depth we're checking at
        """
        if model is None:
            # if the model happens to be None, there is nothing
            # we can say about the validity of the received value
            return True
        elif data is None:
            # assume that means 'empty'
            if type(model).__name__ in ['dict', 'list', 'int', 'float']:
                data = type(model)()
            else:
                data = ''
        if isinstance(model, dict):
            if not isinstance(data, dict):
                self._set_warning(prefix, "Expected dict, "
                                  "gotten a {}".format(type(data).__name__))
                return True
            missing_keys = []
            for k, v in model.items():
                if k not in data:
                    missing_keys.append(k)
                    continue
                p = prefix + '/' + k
                d = data.get(k, None)
                self._check_json_chunk(d, v, prefix=p)
            if len(missing_keys) > 0:
                self._set_warning(prefix,
                                  "Missing keys: {}".format(missing_keys))
        elif isinstance(model, list):
            if not isinstance(data, list):
                self._set_warning(prefix, "Expected list, "
                                  "gotten a {}".format(type(data).__name__))
                return True
            elif len(model) == 1 and len(data) > 1:
                for i in range(len(data)):
                    p = prefix + '[%d]' % i
                    self._check_json_chunk(data[i], model[0], prefix=p)
                return True
            elif len(data) != len(model):
                self._set_warning(prefix, "Expected {} array elements, "
                                  "gotten {}".format(len(model), len(data)))
                return True
            for i in range(len(model)):
                p = prefix + '[%d]' % i
                self._check_json_chunk(data[i], model[i], prefix=p)
        else:
            check = self._verify(model)
            if not check(str(data)):
                self._set_warning(prefix, data)
        return True


class TemplateUrl(object):

    """
    A very partial implementation of RFC 6570, limited to our use
    """
    transforms = {
        'simple': lambda x: x,
        'optional': lambda x: '/' + x,
        'multiple': lambda x: '/'.join(x)
    }
    default = {}
    base = re.compile('(\{.+?\})', re.U)

    def __init__(self, url_string):
        """
        Initialize the template

        Args:
            url_string (str): The url template
        """
        logging.info('Using TemplateUrl, which violates swagger/openapi specification')
        Token = namedtuple('Token', ['key', 'types', 'original'])
        self._url_string = url_string
        self.tokens = []
        for param in self.base.findall(self._url_string):
            types = ['simple']
            key = param.strip('{}')
            if key.startswith('/'):
                types.append('optional')
                key = key.lstrip('/')
            if key.startswith('+'):
                types.append('multiple')
                key = key.lstrip('+')
            self.tokens.append(Token(original=param, key=key, types=types))

    def realize(self, params):
        """
        Returns an url based on the template.

        Args:
            params (dict): the list of params to substitute in the template
        """
        realized = self._url_string
        p = {}
        p.update(self.default)
        p.update(params)
        for token in self.tokens:
            if token.key in p:
                v = p[token.key]
                if isinstance(v, list):
                    v = map(quote_plus, map(str, v))
                else:
                    v = quote_plus(str(v))
                for transform in reversed(token.types):
                    v = self.transforms[transform](v)
            else:
                v = u""
            realized = realized.replace(
                token.original, v, 1)

        return realized


def main():
    parser = argparse.ArgumentParser(
        description='Checks the availability and response of one WMF service')
    parser.add_argument('host_ip', help="The IP address of the host to check")
    parser.add_argument('service_url',
                        help="The base url for the service, including port")
    parser.add_argument('-t', dest="timeout", default=5, type=int,
                        help="Timeout (in seconds) for each "
                        "request. Default: 5")
    parser.add_argument('-s', dest="spec_url", default="/?spec",
                        help="Specific spec url relative to the base one."
                        " Defaults to /?spec.")
    parser.add_argument('-k', dest="insecure",
                        action='store_true',
                        default=False,
                        help="Allow insecure server connections when using SSL."
                        " Defaults to False.")
    args = parser.parse_args()
    metrics_manager = StatsDMetrics(
        host=os.environ.get('STATSD_HOST', default=None),
        port=os.environ.get('STATSD_PORT', default=None),
        prefix=os.environ.get('STATSD_PREFIX', default=None)
    )
    checker = CheckService(args.host_ip, args.service_url,
                           args.timeout, args.spec_url, metrics_manager,
                           args.insecure)
    checker.run()


if __name__ == '__main__':
    main()
